### PR TITLE
Split to a Grafana Service

### LINKED ISSUE #177: Let's do some big refactoring
So, I'm currently working at a feature where I need to respond with multiple dashboards from multiple places. I don't have to option to just create that dashboard and return it. The "easiest way" to implement the feature is to do some rewriting of the package into 4 components: the bot, an adapter, a service and a client.

```txt
┌────────────────────────────┐   ┌────────────────────────┐    ┌────────────────────────┐ 
│ BOT                        │   │ ADAPTER                │    │ CUSTOM RESPONDER       │ 
│                            │   │                        │    │                        │ 
│ handles request from Hubot ├───► respond to a specific  ├────► influence the way the  │ 
│ responsible for response   │   │ chat platform          │    │ response is written    │ 
│ formatting                 │   │                        │    │                        │ 
└──────────────┬─────────────┘   └────────────────────────┘    └────────────────────────┘ 
               │                                                                          
┌──────────────▼─────────────┐                                                            
│ SERVICE                    │                                                            
│                            │                                                            
│ business rules for         │                                                            
│ querying of dashboards     │                                                            
└──────────────┬─────────────┘                                                            
               │                                                                          
┌──────────────▼─────────────┐                                                            
│  CLIENT                    │                                                            
│                            │                                                            
│  interaction with the      │                                                            
│  Grafana API over HTTP     │                                                            
└────────────────────────────┘                                                                          
```

The service should return small DTOs so we can build these new types of integrations.

**Way forward**

I think we should do the following:

- Check if we can refactor the `GrafanaClient` as not to use Hubot HTTP, but something regular like `fetch`. We've tried in the past, but could not get it to work because of the testing (if I remember correctly).
- Check if we can refactor the a `GrafanaService` out of the bot that interacts with the bot based on `DTOs`.
- Find a solutions to `GrafanaClient` creation. Sometimes the client uses data from the channel. This might be fixed by introducing a `GrafanaClientFactory` that will help to determine the right settings for the client.
- With set `setResponder` it is easier for users to influence the way the dashboard is formatted (so we don't need to use middelware like https://keestalkstech.com/2022/11/grafana-i-want-slugs-back-for-my-hubot-%f0%9f%98%ad/#make-hubot-understand anymore.

### PR DESCRIPTION
**Pull request by KeesCBakker**
Partially fixes #177 in which we do a refactor to:

```txt
┌────────────────────────────┐   ┌────────────────────────┐    ┌────────────────────────┐ 
│ BOT                        │   │ ADAPTER                │    │ CUSTOM RESPONDER       │ 
│                            │   │                        │    │                        │ 
│ handles request from Hubot ├───► respond to a specific  ├────► influence the way the  │ 
│ responsible for response   │   │ chat platform          │    │ response is written    │ 
│ formatting                 │   │                        │    │                        │ 
└──────────────┬─────────────┘   └────────────────────────┘    └────────────────────────┘ 
               │                                                                          
┌──────────────▼─────────────┐                                                            
│ SERVICE                    │                                                            
│                            │                                                            
│ business rules for         │                                                            
│ querying of dashboards     │                                                            
└──────────────┬─────────────┘                                                            
               │                                                                          
┌──────────────▼─────────────┐                                                            
│  CLIENT                    │                                                            
│                            │                                                            
│  interaction with the      │                                                            
│  Grafana API over HTTP     │                                                            
└────────────────────────────┘
```

So in this PR we are doing the following:

- Combining `Adapter`, `GrafanaService` and the `GrafanaClient` into a `Bot` class. This makes it a bit easier to interact with the package and not having to build your own class. I could not get the new `Bot` class to handle all the Hubot requests.
- Improved failing test speeds. Tests should be finished withing 100ms. This was 2000ms, causing failing tests to run way too long for refactoring. Also: `--forbid-only --forbid-pending` when doing a commit or push, so we fail when we try to commit a `mocka.only`.
- Improved test coverage by adding tests for pausing and unpausing all alerts.
- The `GrafanaService` has all the business rules. One could use them without the bot (to collect multiple dashboards before sending them on).
- Added some types to `types.d.ts` so JSDoc understands what we're doing.
- With set setResponder it is easier for users to influence the way the dashboard is formatted (so we don't need to use middleware like https://keestalkstech.com/2023/08/improving-responsiveness-of-hubot-grafana/#putting-it-all-together anymore.

### CODE REVIEW

**Comment by KeesCBakker** [comment]
Interesting coverage. We might need to add some extra tests.

image

**Comment by KeesCBakker** [comment]
"It's not easy being green!" 😁

image

File: .husky/pre-commit [hunk 1]
```diff
@@ -1,4 +1,4 @@
 #!/bin/sh
 . "$(dirname "$0")/_/husky.sh"
 
-npm test
+npm test -- --forbid-only --forbid-pending
```
Comment (stephenyeargin) [inline .husky/pre-commit:4]: Today I learned `--forbid-only`. That's a handy one to have.

File: package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "hubot-grafana",
   "description": "Query Grafana dashboards",
-  "version": "5.1.2",
+  "version": "6.0.0",
```
Comment (stephenyeargin) [inline package.json:4]: Let's update the generated release notes with any breaking backwards changes.

File: src/adapters/Adapter.js [hunk 3]
```diff
@@ -80,7 +80,7 @@ class Adapter {
       case 'slack':
         return new SlackUploader(this.robot, this.robot.logger);
       case 'telegram':
-        return new TelegramUploader()
+        return new TelegramUploader();
```
Comment (stephenyeargin) [inline src/adapters/Adapter.js:83]: Don't have to do it for this, but we should probably add `eslint --fix` + [`lint-staged`](https://github.com/lint-staged/lint-staged). My day job's repos are wired up for it, and I've found it to be a time saver.

**Review by stephenyeargin — APPROVED** [review]
Haven't had a chance to load it up on to a running Hubot, but the code looks great.

**Comment by KeesCBakker** [comment]
Will test it on a live Hubot today. Have a great new feature I want to add to our bot, so at the end of the day we'll know.

File: package.json (hunk 2 above)
Comment (KeesCBakker) [reply package.json:4]: Yeah, I'm not 100% that it is really breaking. But I'm also not 100% sure it is not breaking, as we moved some files around. So if you are importing stuff from deep in the package, stuff might break.

File: src/adapters/Adapter.js (hunk 3 above)
Comment (KeesCBakker) [reply src/adapters/Adapter.js:83]: ESLint is super broken in the dev container. It says something about `sinon` being required. When I "fixed" that, it gave me so much errors, that I had to hide under the desk. So we might need to figure out how to let `eslint` play nice.

Comment (KeesCBakker) [reply src/adapters/Adapter.js:83]: In most of my projects I use Prettier (https://prettier.io/docs/en/rationale.html) and just format before committing. Example: https://github.com/KeesCBakker/hubot-command-mapper/blob/master/package.json#L16

**Comment by KeesCBakker** [comment]
Did some testing and some further refactoring. All my tests with a live Hubot were also OK.

image

@stephenyeargin, if it is up to me we can start a merge.

**Review by stephenyeargin — APPROVED** [review]

**Comment by stephenyeargin** [comment]
Looks good to me, I'll let you do the honors. I see you're sneaking it a bit of TypeScript. 😉

**Comment by KeesCBakker** [comment]
> Looks good to me, I'll let you do the honors. I see you're sneaking it a bit of TypeScript. 😉

Hahaha. Yeah the d.ts files work great together with JS Docs.

**Pull request merged**